lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,355 @@
+          }
+        }
+    }
+    if(_ruleStartPoints.length == originalRuleSets.length ) { // All rules are to be deleted
+      _removeDirective(removals, usedDirective);
+    }
+    else {
+      _removeStoredRules(removals, _ruleStartPoints, _ruleEndPoints);
+    }
+  }
+
+  _storeRemovableDeclarations(_PendingRemovals removals, var originalTopLevels, var flippedTopLevels, RetensionMode mode, int iTopLevel) {
+    var originalDecls = originalTopLevels[iTopLevel].declarationGroup.declarations;
+    var flippedDecls = flippedTopLevels[iTopLevel].declarationGroup.declarations;
+
+    /// Iterate over Declarations in RuleSet and store start and end points of declarations to be removed.
+    for (int iDecl = 0; iDecl < originalDecls.length; iDecl++) {
+      if (originalDecls[iDecl] is Declaration && flippedDecls[iDecl] is Declaration) {
+        if (shouldRemoveDecl(mode, originalDecls[iDecl], flippedDecls[iDecl])) {
+          var decls = mode == RetensionMode.keepFlippedBidiSpecific ? flippedDecls : originalDecls;
+          removals.addDeclRemoval(decls[iDecl].span.start.offset, _getDeclarationEnd(removals.source, iDecl, originalDecls, decls));
+        }
+      }
+      else {
+        checkState(originalTopLevels[iTopLevel].runtimeType == flippedTopLevels[iTopLevel].runtimeType);
+      }
+    }
+  }
+
+  _appendDirectionToRuleSet(_PendingRemovals removals, RuleSet ruleSet, Direction targetDirection) {
+    removals._transaction.edit(ruleSet.span.start.offset, ruleSet.span.end.offset,
+        ':host-context([dir="${enumName(targetDirection)}"]) ' +
+            ruleSet.span.text);
+  }
+
+  _removeStoredRules(_PendingRemovals removals, List<int> ruleStartPoints, List<int> ruleEndPoints) {
+    for(int iPoint = 0; iPoint < ruleStartPoints.length; iPoint++) {
+      removals._transaction.edit(ruleStartPoints[iPoint], ruleEndPoints[iPoint], '');
+    }
+  }
+  /// Removes a rule from the transaction.
+  _removeRuleSet(_PendingRemovals removals, int iTopLevel) {
+    removals._transaction.edit(_getRuleSetStart(removals._topLevels[iTopLevel]), _getRuleSetEnd(removals._topLevels, iTopLevel, removals.source.length), '');
+  }
+
+  _removeDirective(_PendingRemovals removals, TreeNode topLevel) {
+    removals._transaction.edit(_getRuleSetStart(topLevel) , topLevel.span.end.offset, '');
+  }
+
+  static int _getRuleSetStart(TreeNode node) {
+    if(node is RuleSet)
+      return node.span.start.offset;
+    // In case of Directives since the node span start does not include '@' so additional -1 is required.
+    return node.span.start.offset - 1;
+  }
+
+  static int _getRuleSetEnd(List<RuleSet> ruleSets, int iTopLevel, int parentEnd) {
+    var end = iTopLevel < ruleSets.length - 1
+        ? _getRuleSetStart(ruleSets[iTopLevel + 1])
+        : parentEnd;
+    return end;
+  }
+
+  static int _getDeclarationEnd(String source, int iDecl, var originalDecls, var decls) {
```

Comment:
Please *never* accept var parameters: it's impossible to read this signature and understand what's going on. This goes for the whole file.